### DISCUSSION TITLE
Verifying JWT

### DISCUSSION THREAD

**Discussion by JMKJR** [nextauthjs/next-auth#2409] [Help] [answered]
I'm using the `jsonwebtoken` library to verify my next-auth generated JWTs on a different server. My `jwt` config object in `[...nextauth].js` just has the `secret` value set to the output of `openssl rand -base64 172 | tr -d '\n'` (random bytes). I try to use this same `secret` as the 2nd argument of the `jsonwebtoken.verify` function along with my JWT session token as the 1st argument. I always get an `invalid signature` error when verifying.

I have confirmed:
1. The JWT is not malformed, that's not the issue (there is a separate error code for this, and i've logged my JWT to confirm it matches the one in my next.js app).
2. I logged the `secret` value from both applications and they matched.
3. I specify `"HS512"` as the algorithm for `jsonwebtoken` since that is what is used by next-auth.

`jsonwebtoken` verify call: `jsonwebtoken.verify(jwt, "secret here", {algorithms: ["HS512"]})`

Any idea why I can't seem to validate the signature if the token, secret, and algorithm is the same? Thanks.

**Comment by balazsorban44** [comment]
I'm not sure about the above issue, but I would rather send the access token generated by the OAuth provider or similar for external APIs. Depends on what provider you use though. In the app's API route, you can decode the token with `import { getToken } from "next-auth/jwt"` passing the same secret from the env var to it.

Here is the `getToken` implementation, it may help:

https://github.com/nextauthjs/next-auth/blob/75ca097ff7585016264f4c6848e58da74c77aaf8/src/lib/jwt.js#L97-L143

**Reply by joswayski** [reply]
Wanted to add some extra context as I came across this issue as well.

So to clarify, the JWT's work fine when created by next-auth, never had any issues with it. 

I wanted to create a token for my unit tests on an API routes, so I used the `jsonwebtoken` library to create one for a simulated user and passed it in the `Authorization` header. 
 
EDIT: Stumbled upon your other convo https://github.com/nextauthjs/next-auth/pull/1066 and just decided to use the custom decode function.

If anyone comes across this in the future, below is how to use the custom decode function with the `jsonwebtoken` library. My middleware for the API looks like this because its 4am and dont want to deal with the types right now:

```  
// Use the next-auth .getToken to retrieve the token, whether in cookies OR a in auth headers 
      const tokenRaw: any = await jwt.getToken({ req, secret, raw: true });
      const params = {
        token: tokenRaw,
        secret: secret,
      };

// Call our 'custom' decode token from the jsonwebtoken library
      const token = await Decode(params);

```

Which looks like this:

```
const jwt = require("jsonwebtoken");

export default async ({ token, secret }: any) => {
  return await jwt.verify(token, secret, { algorithms: ["HS512"] });
};
```

And in [...nextAuth]
```
  jwt: {
    secret: process.env.JWT_SECRET,
    // TODO fix types
    encode: async ({ token, secret }: any) => Encode({ token, secret }),
    decode: async ({ token, secret }: any) => Decode({ token, secret }),
  },
```

**Reply by JMKJR** [reply]
This line `return await jwt.verify(token, secret, { algorithms: ["HS512"] });` worked for you? I tried the same but was getting invalid signature.

I was able to get `jose` working but I don't have the time to post the details right now. Will come back later and post how to do it with `jose` in case that helps anyone.

**Reply by joswayski** [reply]
Yep, thats currently working: https://github.com/joswayski/OpenATS/commit/ef6ed7ea2de08da1d893c95b16b958b1ba26ac21

`middleware/withAuthorizer.ts` is where the decoding happens on the API call and where I was getting the signature issue.

**Reply by JMKJR** [reply]
Thanks for sharing. Aren't your `Encode` and `Decode` handlers just performing the default behavior `next-auth` provides if you didn't specify those handlers? Does forcing `next-auth` to use `jsonwebtoken` for `Encode` somehow fix this `invalid signature` issue we were both seeing? Even though I got it working with `jose` i'm still confused how `jwt.verify(token, secret, { algorithms: ["HS512"] });` worked for you given that my code was identical.

**Reply by joswayski** [reply]
Essentially yeah that forces it to use `jsonwebtoken` instead and that seems to have fixed it. Found [this blog post](https://hasura.io/blog/next-js-jwt-authentication-with-next-auth-and-integration-with-hasura/) which is basically the same code as mine and does the same thing (skip all the Hasura stuff lol)

**Reply by JMKJR** [reply]
Interesting, you'd think as long as `next-auth` is using `HS512` by default then it wouldn't matter what library is used to sign the tokens. But I suppose it does matter unless we were initially doing something wrong ¯\\_(ツ)_/¯ . Regardless, thanks for sharing your solution.

**Comment by JMKJR** [comment]
I'm going to try verifying with [jose](https://www.npmjs.com/package/jose) instead of `jsonwebtoken` and see if I get different results :)

**Reply by JMKJR** [reply]
`jose` solution:

1. Generate a signing key with `npx jose newkey -s 512 -t oct -a HS512`. Let's call it JWT_SIGNING_KEY.
2. Put JWT_SIGNING_KEY in your `.env.local` and add to your `[...nextauth].js` `jwt` config. You don't need to set a `secret`.
3. You can now use this same signing key in a different application to verify the tokens with `jose`. Here is the code...

`await jwtVerify(jwt, key, ​{ algorithms: ["HS512"] } );`

where `jwt` is your token string, `key` is your JWK JSON object you generated, `jwtVerify` comes from the `jose` library.

**Reply by imprakharshukla** [reply]
Hey! Can you please elaborate a bit more on this line:
> Put JWT_SIGNING_KEY in your .env.local and add to your [...nextauth].js jwt config. You don't need to set a secret.

Where should I add it in the jwt config?
Thank you!

**Comment by JMKJR** [comment, marked as answer]
2 answers that work:

https://github.com/nextauthjs/next-auth/discussions/2409#discussioncomment-1036202

https://github.com/nextauthjs/next-auth/discussions/2409#discussioncomment-1039664
